lib/rpmfi.c

```diff
@@ -579,9 +580,9 @@ const unsigned char * rpmfilesFSignature(rpmfiles fi, int ix, size_t *len)
 
     if (fi != NULL && ix >= 0 && ix < rpmfilesFC(fi)) {
 	if (fi->signatures != NULL)
-	    signature = fi->signatures + (fi->signaturelength * ix);
+	    signature = fi->signatures + (fi->signaturemaxlen * ix);
```

Comment:
Is this actually correct, given that the lengths can now vary?

Reply:
Is there a proof that this cannot overflow?

Reply:
I guess this should have just been `signaturelengths[ix]` but somehow got missed. Or I'm missing something (non)obvious here...

Reply:
No, this is an integer overflow with potentially bad consequences.

Reply:
rpmfi.c is likely one big integer overflow if you go looking, the overall assumption is that they don't overflow. Which is why I'm not going to bother with this one either.

Reply:
That doesn't mean we should make it _worse_. And if we _know_ that's an issue, it's probably worth spending some time to fix it all in one go (or fix it as much as reasonably possible).

Reply:
It's not any worse because the assumption is the same as with the surrounding code. It's not IDEAL of course but we have this old codebase with million such assumptions and we can't fix it all to address one immediate issue. You don't look for non-optimal welds in the seams first if you have 10m breach in the hull.

Reply:
This one is bad because the IMA signatures come from the unsigned signature header.

Reply:
That said a limit on signature lengths would be another approach.

Reply:
The header is size-capped to 256MB, there's a limit for you.

Reply:
...and here's another: #1858

Reply:
> Is this actually correct, given that the lengths can now vary?

So @dmnks actually did point out the flaw in this patch but it got lost in all the noise.

Reply:
...except that this isn't the flaw, only pointing to the general direction. The *real* flaw is in the part of hex2bin() that's not touched in this patch so it's hard to spot.